wpilibj/src/main/java/edu/wpi/first/wpilibj/frc2/command/CommandScheduler.java

```diff
@@ -0,0 +1,542 @@
+   * nothing if the command is already scheduled. If a command's requirements are not available, it
+   * will only be started if all the commands currently using those requirements have been scheduled
+   * as interruptible.  If this is the case, they will be interrupted and the command will be
+   * scheduled.
+   *
+   * @param interruptible whether the commands should be interruptible
+   * @param commands      the commands to schedule
+   */
+  public synchronized void queueSchedule(boolean interruptible, Command... commands) {
+    synchronized (m_lock) {
+      for (Command command : commands) {
+        m_toSchedule.put(command, interruptible);
+      }
+    }
+  }
+
+  /**
+   * Schedules multiple commands for execution the next time the scheduler is run, with
+   * interruptible defaulted to true.  Thread-safe.  Does nothing if the command is already
+   * scheduled.
+   *
+   * @param commands the commands to schedule
+   */
+  public void queueSchedule(Command... commands) {
+    queueSchedule(true, commands);
+  }
+
+  /**
+   * Runs a single iteration of the scheduler.  The execution occurs in the following order:
+   *
+   * <p>Subsystem periodic methods are called.
+   *
+   * <p>Button bindings are polled, and new commands are scheduled from them.
+   *
+   * <p>Currently-scheduled commands are executed.
+   *
+   * <p>End conditions are checked on currently-scheduled commands, and commands that are finished
+   * have their end methods called and are removed.
+   *
+   * <p>Any subsystems not being used as requirements have their default methods started.
+   */
+  @SuppressWarnings({"PMD.CyclomaticComplexity", "PMD.NPathComplexity"})
+  public void run() {
+    if (m_disabled) {
+      return;
+    }
+
+    //Run the periodic method of all registered subsystems.
+    for (Subsystem subsystem : m_subsystems.keySet()) {
+      subsystem.periodic();
+    }
+
+    //Poll buttons for new commands to add.
+    for (Runnable button : m_buttons) {
+      button.run();
+    }
+
+    synchronized (m_lock) {
+      //Add commands from async queue
+      for (Map.Entry<Command, Boolean> command : m_toSchedule.entrySet()) {
+        schedule(command.getValue(), command.getKey());
+      }
+      m_toSchedule.clear();
+      //Cancel commands from async queue
+      cancel(m_toCancel.toArray(new Command[]{}));
+      m_toCancel.clear();
+    }
+
+    //Run scheduled commands, remove finished commands.
+    for (Iterator<Command> iterator = m_scheduledCommands.keySet().iterator();
+         iterator.hasNext(); ) {
+      Command command = iterator.next();
+
+      if (!command.runsWhenDisabled() && RobotState.isDisabled()) {
+        command.end(true);
+        for (Consumer<Command> action : m_interruptActions) {
+          action.accept(command);
+        }
+        m_requirements.keySet().removeAll(command.getRequirements());
+        iterator.remove();
```

Comment:
Does iterator.remove() remove command, or the next command, since it was advanced with next()?

Reply:
It removes command; next returns the element after incrementing.